src/Utils/BlockTemplateUtils.php

```diff
@@ -244,15 +280,65 @@ public static function generate_template_slug_from_path( $path, $directory_name
 		);
 	}
 
+	/**
+	 * Gets the first matching template part within themes directories
+	 *
+	 * Since [Gutenberg 12.1.0](https://github.com/WordPress/gutenberg/releases/tag/v12.1.0), the conventions for
+	 * block templates and parts directory has changed from `block-templates` and `block-templates-parts`
+	 * to `templates` and `parts` respectively.
+	 *
+	 * This function traverses all possible combinations of directory paths where a template or part
+	 * could be located and returns the first one which is readable, prioritizing the new convention
+	 * over the deprecated one, but maintaining that one for backwards compatibility.
+	 *
+	 * @param string $template_slug  The slug of the template (i.e. without the file extension).
+	 * @param string $template_type  Either `wp_template` or `wp_template_part`.
+	 *
+	 * @return string|null  The matched path or `null` if no match was found.
+	 */
+	public static function get_theme_template_path( $template_slug, $template_type = 'wp_template' ) {
+		$template_filename      = $template_slug . '.html';
+		$possible_templates_dir = 'wp_template' === $template_type ? array(
+			self::TEMPLATES_DIR_NAME,
+			self::DEPRECATED_TEMPLATES_DIR_NAME,
+		) : array(
+			self::TEMPLATE_PARTS_DIR_NAME,
+			self::DEPRECATED_TEMPLATE_PARTS_DIR_NAME,
+		);
+
+		// Combine the possible root directory names with either the template directory
+		// or the stylesheet directory for child themes.
+		$possible_paths = array_reduce(
```

Comment:
In my tests, at this point the array has 4 entries when it should have 2 I think? There are duplicates.

Reply:
Ah, I see why now. This will also consider child themes. We should probably check the value doesnt exist in the array before adding it.

Reply:
`$possible_paths` here should be 4 entries, because it combines the two possible conventions (newer and deprecated) and the two possible roots (`template_directory` and `stylesheet_directory`).

For most themes, I suppose, these would be duplicates as `get_template_directory` and `get_stylesheet_directory` return the same path. However, this would allow for child themes, where, AFAIU, these two functions return different paths (the former returns the parent theme's path, the latter the child theme's one)

Reply:
> Ah, I see why now. This will also consider child themes. We should probably check the value doesnt exist in the array before adding it.

I see your point, but I'm not sure this is required. The next bit of logic iterates through the array and returns the first available path. There is basically no harm done (as the array is also super small), and it keeps the code a bit more readable, IMHO.

But if you feel strongly about this, I can add the check.

Reply:
This is fine then, I can see now why it works that way. I agree lets leave it 🙂